src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,309 @@
+	static $eagerness_allowlist = array(
+		'immediate'    => true,
+		'eager'        => true,
+		'moderate'     => true,
+		'conservative' => true,
+	);
+	return isset( $eagerness_allowlist[ $eagerness ] );
+}
+
+/**
+ * Checks whether the given speculation rules source is valid.
+ *
+ * @since 6.8.0
+ *
+ * @param string $source Speculation rules source.
+ * @return bool True if valid, false otherwise.
+ */
+function wp_is_valid_speculation_rules_source( string $source ): bool {
+	static $source_allowlist = array(
+		'list'     => true,
+		'document' => true,
+	);
+	return isset( $source_allowlist[ $source ] );
+}
+
+/**
+ * Returns the full speculation rules data based on the given configuration.
+ *
+ * Plugins with features that rely on frontend URLs to exclude from prefetching or prerendering should use the
+ * {@see 'wp_speculation_rules_href_exclude_paths'} filter to ensure those URL patterns are excluded.
+ *
+ * @since 6.8.0
+ * @access private
+ *
+ * @param array<string, string> $configuration Associative array with 'mode' and 'eagerness' keys.
+ * @return WP_Speculation_Rules Object representing the speculation rules to use.
+ */
+function wp_get_speculation_rules( array $configuration ): WP_Speculation_Rules {
+	// If the passed configuration is invalid, trigger a warning and fall back to the default configuration.
+	if (
+		! isset( $configuration['mode'] ) ||
+		! wp_is_valid_speculation_rules_mode( $configuration['mode'] ) ||
+		! isset( $configuration['eagerness'] ) ||
+		! wp_is_valid_speculation_rules_eagerness( $configuration['eagerness'] )
+	) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			esc_html(
+				sprintf(
+					/* translators: %s is $configuration */
+					__( 'The %s parameter was provided with an invalid value.' ),
+					'$configuration'
+				)
+			),
+			'6.8.0'
+		);
+		$configuration = wp_get_speculation_rules_configuration();
+		if ( null === $configuration ) {
+			return array();
+		}
+	}
+
+	$mode      = $configuration['mode'];
+	$eagerness = $configuration['eagerness'];
+
+	$prefixer = new WP_URL_Pattern_Prefixer();
+
+	$base_href_exclude_paths = array(
+		$prefixer->prefix_path_pattern( '/wp-*.php', 'site' ),
+		$prefixer->prefix_path_pattern( '/wp-admin/*', 'site' ),
+		$prefixer->prefix_path_pattern( '/*', 'uploads' ),
+		$prefixer->prefix_path_pattern( '/*', 'content' ),
+		$prefixer->prefix_path_pattern( '/*', 'plugins' ),
+		$prefixer->prefix_path_pattern( '/*', 'template' ),
+		$prefixer->prefix_path_pattern( '/*', 'stylesheet' ),
+	);
+
+	/*
+	 * If pretty permalinks are enabled, exclude any URLs with query parameters.
+	 * Otherwise, exclude specifically the URLs with a `_wpnonce` query parameter.
+	 */
+	if ( get_option( 'permalink_structure' ) ) {
+		$base_href_exclude_paths[] = $prefixer->prefix_path_pattern( '/*\\?(.+)', 'home' );
+	} else {
+		$base_href_exclude_paths[] = $prefixer->prefix_path_pattern( '/*\\?*(^|&)_wpnonce=*', 'home' );
```

Comment:
Should this be more general? For example, should we exclude the `_wp` part and match any query parameter that contains `nonce`? That could catch more instances where third plugins have action-based URLs to avoid prefetching/prerendering them.

Reply:
That's a fair point. We should probably also look for the name `nonce` on its own, or anything containing `_nonce`. Worth noting that this is only relevant for the small fraction of WordPress sites that don't use pretty permalinks, but still worth supporting probably.

@westonruter Would you mind amending this URL pattern? I think you're more familiar with how these work than myself.

Reply:
The reason for `_wpnonce` is that this is specifically what WordPress uses by default in `wp_nonce_url()`.  If we want to match any query parameter that contains `nonce` then this would be the pattern:

```suggestion
		$base_href_exclude_paths[] = $prefixer->prefix_path_pattern( '/*\\?*(^|&)*nonce*=*', 'home' );
```

Examples:

* [`nonce=...`](https://urlpattern.com/#eyJ1cmxQYXR0ZXJuIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8qXFw/KihefCYpKm5vbmNlKj0qIiwidXJsIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8/bm9uY2U9YXNkYXNkIn0=)
* [`_wpnonce=...`](https://urlpattern.com/#eyJ1cmxQYXR0ZXJuIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8qXFw/KihefCYpKm5vbmNlKj0qIiwidXJsIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8/X3dwbm9uY2U9YXNkYXNkIn0=)
* [`noncewp=...`](https://urlpattern.com/#eyJ1cmxQYXR0ZXJuIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8qXFw/KihefCYpKm5vbmNlKj0qIiwidXJsIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8/bm9uY2V3cD1hc2Rhc2QifQ==)
* [`...&nonce=...`](https://urlpattern.com/#eyJ1cmxQYXR0ZXJuIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8qXFw/KihefCYpKm5vbmNlKj0qIiwidXJsIjoiaHR0cHM6Ly9leGFtcGxlLmNvbS8/Zm9vPWJheiZub25jZT1mb28ifQ==)

Reply:
I don't feel strongly on this. But it was something I thought of when thinking through scenarios where the feature could unintentionally cause undesired behavior. It's also only for sites without pretty permalinks enabled, which is likely a small subset of sites.

Reply:
Just updated this in https://github.com/WordPress/wordpress-develop/pull/7860/commits/b08592a5c75c16b84fd3479bfdcb0238ac11efe1.

Probably not a big deal, but there's no harm in adding this extra safeguard.